airflow/models/dag.py

```diff
@@ -1467,22 +1467,26 @@ def _test_cycle_helper(self, visit_map, task_id):
         """
         from airflow.models.dagbag import DagBag  # Avoid circular imports
 
-        # print('Inspecting %s' % task_id)
-        if visit_map[task_id] == DagBag.CYCLE_DONE:
-            return False
-
-        visit_map[task_id] = DagBag.CYCLE_IN_PROGRESS
-
-        task = self.task_dict[task_id]
-        for descendant_id in task.get_direct_relative_ids():
-            if visit_map[descendant_id] == DagBag.CYCLE_IN_PROGRESS:
-                msg = "Cycle detected in DAG. Faulty task: {0} to {1}".format(
-                    task_id, descendant_id)
+        path, visited = [task_id], []
+
+        while path:
+            current_task_id = path.pop()
+            if visit_map[current_task_id] == DagBag.CYCLE_DONE:
+                return False
+            if visit_map[current_task_id] == DagBag.CYCLE_IN_PROGRESS:
+                msg = "Cycle detected in DAG. Faulty task: {}".format(
+                    current_task_id)
                 raise AirflowDagCycleException(msg)
-            else:
-                self._test_cycle_helper(visit_map, descendant_id)
-
-        visit_map[task_id] = DagBag.CYCLE_DONE
+            visit_map[current_task_id] = DagBag.CYCLE_IN_PROGRESS
+            task = self.task_dict[current_task_id]
+            for descendant_id in task.get_direct_relative_ids():
+                path.append(descendant_id)
+            if current_task_id in path:
```

Comment:
I'm a bit concerned about this check. `path` is a list and in the context of this very issue we assume thousands of dags. This check is linear. Do I miss something? If not, I would suggest benchmarking it against a different structure (set comes to mind)

Reply:
Sounds like a good idea to me (and I'm always in favour of using numbers to make performance decisions)

Reply:
I ran this test on 100,000 tasks and it saved about 1/2 of a second. Don't really think it's worth it.

Reply:
Also this is only run on one DAG at a time, so the max length is the max # of tasks a user would put into a single DAG.